File: Parser.py

import Words
import Grammar
import logging
import re   # Regex

from Grammar import isAccessModifier

logging.basicConfig(format='[%(filename)s:%(lineno)d] %(message)s', level=logging.DEBUG)
debug = logging.debug

def isAtom(string):
    return re.search("^([a-zA-Z_][a-zA-Z0-9_]*)$", string)




class ExpressionWithIndentation:
    def __init__(self, expression, indentation):
        self.expression = expression
        self.indentation = indentation
    def __str__(self):
        return ' '*self.indentation + str(self.expression)

class Expression:
    def __init__(self, parent, content, type):
        self.content = content
        self.type = type
        self.parent = parent
        self.accessModifiers = []
    def __str__(self):
        if True:
            pass
        else:
            logging.error('ERROR: Type ' + self.type + ' not handled!')

class Node:
    def __init__(self, content, type=None):
        self.content = content
        self.type = type
    def __str__(self):
        return str(self.content)






class Parser:
    def __init__(self, wordLine):
        self.wordLine = wordLine
        self.root = Expression(content=[], type='expression', parent=None)
        self.currentExpression = self.root
        self.stateStack = ['in-root']

    def exit(self, message):
        logging.error(message)

    def push(self, what):
        self.currentExpression.content.append(what)
        logging.debug('Pushing ' + str(what.content))

    def getCurrentState(self):
        return self.stateStack[-1]
    def setState(self, newState):
        self.stateStack[-1] = newState

    def branchOut(self, expressionType, newState=None):
        newExpression = Expression(content=[], type=expressionType, parent=self.currentExpression)
        self.push(newExpression)
        self.currentExpression = newExpression
        if newState is not None:
            self.stateStack.append(newState)

    def brateIn(self):
        self.stateStack.pop()
        self.currentExpression = self.currentExpression.parent

    def wrapOver(self, newExpressionType, nextState):  # Wraps the current expression in another expression
        accessModifiers = self.currentExpression.accessModifiers
        content = self.currentExpression.content
        self.currentExpression.accessModifiers = []
        self.currentExpression.content = []
        newExpression = Expression(self.currentExpression, content, newExpressionType)
        newExpression.accessModifiers = accessModifiers
        self.currentExpression.content = [newExpression]
        self.currentExpression = newExpression
        self.stateStack.append(nextState)

    def parse(self):
        strings = [word.string for word in self.wordLine.words]
        stateFunctions = {
            'no-state'                      : self.noState,
            'in-root'                       : self.inRoot
        }

        for word in strings:
            state = self.getCurrentState()
            if state in stateFunctions:
                stateFunctions[state](word)
            else:
                logging.debug('Error: State ' + state + ' not handled.')
            logging.debug('> ' + str(self.root) + '\t\t|\t' + state)
        return self.root







    def noState(self, string):
        self.exit('No state for string ' + string)
            
    def inRoot(self, string):
        if isAccessModifier(string):
            self.currentExpression.accessModifiers.append(string)
            self.setState('reading-modifiers')

    def readingModifiers(self, string):
        if isAccessModifier(string):
            self.currentExpression.accessModifiers.append(string)
        elif isAtom(string):
            self.readingType(string)
            
    def readingType(self, string):
        if isAtom(string):
            self.push(Node(string))
            self.setState('expecting-generic')
    
    def expectingGeneric(self, string):
        if string == '<':
            self.branchOut('generic-inner', 'reading-generic-inner')
        elif isAtom(string):
            self.readingVarName(string)
    
    def readingGenericInner(self, string):
        if string == '>':
            self.brateIn()
        elif isAtom(string):
            self.push(Node(string)) # For now, allows < anything anything anything... >

    def readingVarName(self, string):
        if isAtom(string):
            self.push(Node(string))



















    def inParenthesisExpression(self, string):
        if string == '(':
            self.branchOut('expression-or-tuple', 'in-parenthesis-expression')
        elif string == ',':
            self.currentExpression.type = 'tuple'
            self.stateStack[-1] = 'in-parenthesis-tuple'
            thisContent = self.currentExpression.content
            tupleElement = Expression(content=thisContent, type='expression', parent=self.currentExpression)
            self.currentExpression.content = [tupleElement]
            self.branchOut('expression', 'in-tuple-element')
        elif string == ')':
            self.brateIn()
        else:
            self.push(Node(string, 'node'))

    def inParenthesisTuple(self, string):
        logging.error('Not ok')
        pass

    def inTupleElement(self, string):
        if string == '(':
            self.branchOut('expression-or-tuple', 'in-parenthesis-expression')
        elif string == ',':
            self.brateIn()
            self.branchOut('expression','in-tuple-element')
        elif string == ')':
            self.brateIn()
            self.brateIn()
        else:
            self.push(Node(string, 'node'))


def parseWordLines(wordLines):
    return [ExpressionWithIndentation(Parser(wordLine).parse(), wordLine.indentation) for wordLine in wordLines]

def printExpressionWithIndentationList(expressionWithIndentationList):
    for expression in expressionWithIndentationList:
        logging.debug(str(expression))

File: JS/Parser.js

Words = require('./Words')
Grammar = require('./Grammar')
Expression = require('./Expressions').Expression







class Parser {
    constructor(wordLine) {
        this.wordLine = wordLine
        this.root = new Expression(null, [], 'root-expression')
        this.currentExpression = this.root
        this.stateStack = ['in-root']
        this.currentWord = null     // Set in the for in parse
    }

    exit(message) { console.log('Error: ' + message);  throw 'Exiting' }
    error() { throw `Error in state ${this.getCurrentState()} at word ${this.currentWord.string} with type ${this.currentWord.type}` }
    push(what) { this.currentExpression.content.push(what) }
    getCurrentState() { return this.stateStack[this.stateStack.length - 1] }
    setState(newState) { this.stateStack[this.stateStack.length - 1] = newState }

    redirectToState(toState) {
        this.setState(toState)
        this[this.getStateFunction(toState)]()
    }

    branchOut (newExpressionType, newState=null) {  // Goes up 1 level and optionally 1 state
        let newExpression = new Expression(this.currentExpression, [], newExpressionType)
        this.push(newExpression)
        this.currentExpression = newExpression
        if (newState != null) this.stateStack.push(newState)
    }

    brateIn () {    // Goes back 1 state and back 1 level
        this.stateStack.pop()
        this.currentExpression = this.currentExpression.parent
    }

    wrapOver ({wrapperExpressionType, newExpressionType, nextState}) {   // Takes the current expression's content and modifiers, creates a new expression and puts those in the new expression; also changes state
        if (wrapperExpressionType != null)
            this.currentExpression.type = wrapperExpressionType
        let accessModifiers = this.currentExpression.accessModifiers
        let content = this.currentExpression.content
        this.currentExpression.accessModifiers = []
        this.currentExpression.content = []
        let newExpression = new Expression(this.currentExpression, content, newExpressionType)
        newExpression.accessModifiers = accessModifiers
        this.currentExpression.content = [newExpression]
        this.currentExpression = newExpression
        this.stateStack.push(nextState)
        console.log('this.currentExpression')
        console.log(this.currentExpression)
    }

    getStateFunction(stateName) {   // Each state is mapped to a function (don't ask me why they are not just called the same)
        return Utils.dashCaseToCamelCase(stateName)
    }

    parse () {
        for (let word of this.wordLine.words) {
            this.currentWord = word
            let state = this.getCurrentState()
            console.log(`Word: ${word}    State ${state}`)
            let functionName = this.getStateFunction(state)
            if (this[functionName] != null) {
                this[functionName]()
            } else {
                this.exit('State ' + state + ' not handled.')
            }
        }
        return { expression: this.root, indentation: this.wordLine.indentation }
    }

    /* States */

    noState() { this.error() }

    inRoot() {
        switch (this.currentWord.type) {
            case 'MODIFIER':
                this.redirectToState('reading-modifiers')
                break
            default: this.error()
        }
    }

    readingModifiers() {
        switch (this.currentWord.type) {
            case 'MODIFIER':
                this.currentExpression.accessModifiers.push(this.currentWord.string)
                break
            case 'ATOM':
                this.redirectToState('reading-type')
                break
            case 'CLASS':
                this.redirectToState('reading-class-declaration')
                break
            case 'FUNC':
                this.redirectToState('reading-function-declaration')
            default: this.error()
        }
    }

    readingFunctionDeclaration() {
        switch (this.currentWord.type) {
            case 'FUNC':
                this.branchOut('function-declaration', 'expecting-function-generic')
                break
            default: this.error()
        }
    }

    readingClassDeclaration() {
        switch (this.currentWord.type) {
            case 'CLASS':
                this.push(this.currentWord.string)
                this.setState('expecting-class-generic')
                break
            default: this.error()
        }
    }

    expectingFunctionGeneric() {
        switch (this.currentWord.type) {
            case '<':
                this.branchOut('function-generic', 'reading-generic-inner')
                break
            case 'ATOM':
                this.redirectToState('reading-function-name')
                break
            default: this.error()
        }
    }

    expectingClassGeneric() {
        switch(this.currentWord.type) {
            case '<':
                this.branchOut('class-generic', 'reading-generic-inner')
                break
            case 'ATOM':
                this.redirectToState('reading-class-name')
                break
            default: this.error()
        }
    }

    readingType() {
        switch (this.currentWord.type) {
            case 'ATOM':
                this.push(this.currentWord.string)
                this.setState('expecting-generic')
                break
            default: this.error()
        }
    }

    expectingGeneric() {
        switch (this.currentWord.type) {
            case '<':
                this.branchOut('generic-inner', 'reading-generic-inner')
                break
            case 'ATOM':
                this.redirectToState('reading-var-name')
                break
            default: this.error()
        }
    }

    readingGenericInner() {
        switch (this.currentWord.type) {
            case '>':
                this.brateIn()
                break
            case 'ATOM':
                this.push(this.currentWord.string)
                break
            default: this.error()
        }
    }

    readingVarName() {
        switch (this.currentWord.type) {
            case 'ATOM':
                this.push(this.currentWord.string)
                this.setState('expecting-attribution-equals')
                break
            default: this.error()            
        }
    }

    readingClassName() {
        switch (this.currentWord.type) {
            case 'ATOM':
                this.push(this.currentWord.string)
                this.setState('no-state')
                break
            default: this.error()            
        }
    }

    readingFunctionName() {
        switch (this.currentWord.type) {
            case 'ATOM':
                this.push(this.currentWord.string)
                this.setState('no-state')
                break
            default: this.error()            
        }
    }

    expectingAttributionEquals() {
        switch (this.currentWord.type) {
            case '=':
                this.wrapOver({
                    wrapperExpressionType: 'attribution',
                    newExpressionType: this.currentExpression.type,
                    nextState: 'none'   // No need for a new state
                })
                this.brateIn()
                this.branchOut('attribution-right', 'reading-attribution-right')
                break
            default:
                this.error()
        }
    }

    readingAttributionRight() {
        switch (this.currentWord.type) {
            default: this.push(this.currentWord.string)
        }
    }

}

// Takes a list of WordLine
// Returns a list of {expression, indentation}
function parseWordLines(wordLines){
    return wordLines.map( line => (new Parser(line).parse()))
}

module.exports = { Parser : Parser }
__requirer['Parser'] = module.exports
__requirer['./Parser'] = module.exports

